app/src/main/java/com/hedvig/app/feature/genericauth/otpinput/OtpInputViewModel.kt

```diff
@@ -0,0 +1,110 @@
+package com.hedvig.app.feature.genericauth.otpinput
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.hedvig.app.authenticate.AuthenticationTokenService
+import com.hedvig.app.util.ErrorEvent
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class OtpInputViewModel(
+    private val otpId: String,
+    private val credential: String,
+    private val authenticationTokenService: AuthenticationTokenService,
+    private val sendOtpCodeUseCase: SendOtpCodeUseCase,
+    private val reSendOtpCodeUseCase: ReSendOtpCodeUseCase,
+) : ViewModel() {
+    private val _viewState = MutableStateFlow(ViewState(credential = credential))
+    val viewState = _viewState.asStateFlow()
+
+    private val eventChannel = Channel<Event>(Channel.UNLIMITED)
+    val eventsFlow = eventChannel.receiveAsFlow()
+
+    data class ViewState(
+        val input: String = "",
+        val credential: String,
+        val errorMessage: String? = null,
+        val errorEvent: ErrorEvent? = null,
+        val loadingResend: Boolean = false,
+        val loadingCode: Boolean = false
+    )
+
+    sealed class Event {
+        data class Success(val authToken: String) : Event()
+        object ShowDialog : Event()
+        object CodeResent : Event()
+        object None : Event()
```

Comment:
Why do we need None as an event? Is it to clear past events? If that's so, it probably stems from the fact that we're saving the event as a state?

Reply:
Yeah this is ugly. I couldnt find a better solution, see here: https://stackoverflow.com/a/69358848

Reply:
Aha I see. Can you test this with the solution given above? Of receiving the event inside the LaunchedEffect and never saving it in a state therefore not relying on clearing it and setting it again as all unique events should be consumed even if they are the same since they come from a Channel and consumed right away.